### PR TITLE
adding rpc for getting mapping urls

### PR DESCRIPTION
**Pull request by mikedoan**
adding rpc for getting mapping urls

### CODE REVIEW

**Comment by mikedoan** [comment]
jenkins, test this please.

File: internal/hmyapi/blockchain.go [hunk 1]
```diff
@@ -63,6 +64,63 @@ func (s *PublicBlockChainAPI) GetBlockByHash(ctx context.Context, blockHash comm
...
 }
 
+// GetShardingStructure returns an array of sharding structures.
+func (s *PublicBlockChainAPI) GetShardingStructure(ctx context.Context) ([]map[string]interface{}, error) {
+	if core.ShardingSchedule.GetNetworkID() == shardingconfig.MainNet {
+		return []map[string]interface{}{
+			map[string]interface{}{
+				"current": s.b.GetShardID() == 0,
+				"shardID": "0",
+				"http":    "http://s0.t.hmy.io:9500",
```
Comment (LeoHChen) [inline internal/hmyapi/blockchain.go]: hmny.io

File: internal/hmyapi/blockchain.go [hunk 2]
```diff
@@ -63,6 +64,63 @@ func (s *PublicBlockChainAPI) GetBlockByHash(ctx context.Context, blockHash comm
...
+				"ws":      "ws://s0.t.hmy.io:9800",
+			},
+			map[string]interface{}{
+				"current": s.b.GetShardID() == 1,
+				"shardID": "1",
+				"http":    "http://s1.t.hmy.io:9500",
```
Comment (LeoHChen) [inline internal/hmyapi/blockchain.go]: also, please make it more structure, I don't like hardcoded everything here.

File: internal/hmyapi/blockchain.go [hunk 3]
```diff
@@ -63,6 +64,63 @@ func (s *PublicBlockChainAPI) GetBlockByHash(ctx context.Context, blockHash comm
 	return nil, err
 }
 
+// GetShardingStructure returns an array of sharding structures.
+func (s *PublicBlockChainAPI) GetShardingStructure(ctx context.Context) ([]map[string]interface{}, error) {
+	if core.ShardingSchedule.GetNetworkID() == shardingconfig.MainNet {
```
Comment (LeoHChen) [inline internal/hmyapi/blockchain.go]: better wrap this data into the schedule, don't hardcode it here.

File: internal/hmyapi/blockchain.go [hunk 4]
```diff
@@ -63,6 +64,63 @@ func (s *PublicBlockChainAPI) GetBlockByHash(ctx context.Context, blockHash comm
...
+				"http":    "http://s2.t.hmy.io:9500",
+				"ws":      "ws://s2.t.hmy.io:9800",
+			},
+			map[string]interface{}{
+				"shardID": "3",
+				"http":    "http://s3.t.hmy.io:9500",
+				"ws":      "ws://s3.t.hmy.io:9800",
+			},
+		}, nil
+	} else if core.ShardingSchedule.GetNetworkID() == shardingconfig.TestNet {
```
Comment (LeoHChen) [inline internal/hmyapi/blockchain.go]: the string should be generated using a rule, don't hardcode, please.

File: internal/hmyapi/blockchain.go [hunk 5]
```diff
@@ -63,6 +64,63 @@ func (s *PublicBlockChainAPI) GetBlockByHash(ctx context.Context, blockHash comm
...
+				"http":    "http://s0.b.hmy.io:9500",
+				"ws":      "ws://s0.s.hmy.io:9800",
+			},
+			map[string]interface{}{
+				"shardID": "1",
+				"http":    "http://s1.b.hmy.io:9500",
+				"ws":      "ws://s1.s.hmy.io:9800",
+			},
+		}, nil
+	} else {
```
Comment (LeoHChen) [inline internal/hmyapi/blockchain.go]: use case statement, you need to handle pangaea as well, localnet, devnet, all different.

File: internal/hmyapi/blockchain.go (hunk 3 above)
Comment (mikedoan) [reply internal/hmyapi/blockchain.go]: ShardingSchedule is this schedule variable.

File: internal/hmyapi/blockchain.go [hunk 6]
```diff
@@ -63,6 +72,44 @@ func (s *PublicBlockChainAPI) GetBlockByHash(ctx context.Context, blockHash comm
 	return nil, err
 }
 
+// GenShardingStructure return sharding structure, given shard number and its patterns.
+func GenShardingStructure(shardNum, shardID uint32, httpPattern, wsPattern string) []map[string]interface{} {
```
Comment (LeoHChen) [inline internal/hmyapi/blockchain.go]: can you move this function to the schedule?

File: internal/hmyapi/blockchain.go [hunk 7]
```diff
@@ -25,6 +26,14 @@ const (
 	defaultFromAddress = "0x0000000000000000000000000000000000000000"
 )
 
+// Constants for sharding structure.
+const (
+	MainNetHTTPPattern = "http://s%d.t.hmy.io:9500"
+	MainNetWSPattern   = "ws://s%d.t.hmy.io:9800"
+	TestNetHTTPPattern = "http://s%d.s.hmy.io:9500"
```
[the remark this answers is not in the document]
Comment (mikedoan) [reply internal/hmyapi/blockchain.go]: it's confusing here. please clarify

for mainnet: s0.t.hmny.io ?

for testnet: s0.b.hmny.io ?

File: internal/hmyapi/blockchain.go (hunk 6 above)
Comment (mikedoan) [reply internal/hmyapi/blockchain.go]: Do you mean to introduce an interface and implement it for each type of network?

can we do it later as it's the blocker for beta?

**Pull request merged**